enhancements/20240506-cpu-pinning.md

```diff
@@ -0,0 +1,302 @@
+  Currently Harvester use `none` policy which is the default one. And we also need to set kubelet
+  `--kube-reserved` or `--system-reserved` options. This step needs to delete cpu_manager_state file and restart rke2-server/rke2-worker (or you can say kubelet).
+2. Enable KubeVirt `CPUManager` feature gate. Then KubeVirt could add [cputune](https://libvirt.org/formatdomain.html#cpu-tuning)
+  part to libvirt domain xml to pin guest's vCPUs in virtual machine to the host's pCPUs.
+3. Allow virtual machine to apply [Guaranteed QoS](https://kubernetes.io/docs/concepts/workloads/pods/pod-qos/#guaranteed).
+  Currently Harvester force all virtual machines apply overcommitting resources which make all virtual machine pods under
+  [Burstable QoS](https://kubernetes.io/docs/concepts/workloads/pods/pod-qos/#burstable). While the CPU Manager static
+  policy only reserved cpus when both cpu requests and limits values must be the same integer.
+
+In step1, we have to set up the reserved resource due to the k8s cpu manager requirement.
+We can utilize the formula as follows to set `system-reserved` CPU resources during install harvester node. This is inspired by [GKE CPU reservations](https://cloud.google.com/kubernetes-engine/docs/concepts/plan-node-sizes).
+
+```text
+6% of the first core +
+1% of the next core (up to 2 cores) +
+0.5% of the next 2 cores (up to 4 cores) +
+0.25% of any cores above 4 cores
+
+if the maximum number of Pods per node beyond the default of 110 (Currently harvester max pods is 200),
+reserves an extra 400 mCPU in addition to the preceding reservations.
+```
+
+After calculating the cpu reserved resources, add the following content to `/etc/rancher/rke2/config.yaml/d/99-z00-harvester-system-reserved.yaml` during harvester installation.
+
+```yaml
+kubelet-arg+:
+- "system-reserved=cpu=490m"
+```
+
+and add another config `/etc/rancher/rke2/config.yaml/d/99-z01-harvester-cpu-manager.yaml` which includes the cpu-manager-policy setting.
+
+```yaml
+kubelet-arg+:
+- "cpu-manager-policy=none"
+```
+
+> [!NOTE]
+> After upgrading harvester to a newer version, a new file `99-max-pods.yaml` shows up which override all the kubelet-args, that's why I have to 
+> create another two files `/etc/rancher/rke2/config.yaml/d/99-z00-harvester-system-reserved.yaml`, `/etc/rancher/rke2/config.yaml/d/99-z01-harvester-cpu-manager.yaml`
+> instead of adding these configs to `90-harvester-server.yaml` or `90-harvester-worker.yaml`. Another benefit is that if we want to modify cpu manager policy setting,
+> we only need to update one file `99-z01-harvester-cpu-manager.yaml`.
+
+To change cpu-manager policy on the fly, utilizing `Plan` CRD here to deploy jobs to all nodes. We need to modify the secret `cattle-system/cpu-manager`, and fill every policy settings in each node that we want to change in the `.stringdata.configs` field. It should be a json format like the following example.
```

Comment:
Discussed offline with @bk201, we identified two drawbacks with this solution:
- Jobs are sent to all nodes, even though we only need to change the CPU manager policy on specific nodes.
- The process is not straightforward; ideally, there should be a web API instead of requiring developers/users to upload secrets.

I will refine the HEP and redesign how we manage the CPU manager policy per node.